### PR TITLE
WooCommerce: Add variation edit state

### PR DESCRIPTION
**Pull request by coderkevin**
This adds an edit state for product variations which can then be used
for corresponding create/update calls to the API.

To test:

1. Run `make test` and ensure all tests succeed.

### CODE REVIEW

**Comment by allendav** [comment]
Same comment as on #12919 - would it make more sense to conceptualize this as attribute edit state instead of variation state?

**Comment by justinshreve** [comment]
@allendav In #12919 we are dealing with variation types which are stored as attributes. In this case, we are dealing with actual variation objects. 
https://woocommerce.github.io/woocommerce-rest-api-docs/?javascript#product-variations

**Comment by coderkevin** [comment]
I'm not sure that it makes sense here. Variations (http://woocommerce.github.io/woocommerce-rest-api-docs/#product-variations) and Attributes (http://woocommerce.github.io/woocommerce-rest-api-docs/#product-attributes) are two different things to the API. And the API is what we're trying to mirror in the edit state, so we can easily transfer the edit data via the API when the user clicks "Save".

File: client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+function editProductVariationAction( edits, action ) {
+	const { product, variation, data } = action.payload;
+	const prevEdits = edits || [];
+	const productId = product.id;
+	const bucket = variation && isNumber( variation.id ) && 'updates' || 'creates';
+
+	let found = false;
+
+	// Look for an existing product edits first.
+	const _edits = prevEdits.map( ( productEdits ) => {
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js]: Why the leading underscore for `_edits` and `_array`?

File: client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+		if ( productId === productEdits.productId ) {
+			found = true;
+			const _array = editProductVariation( productEdits[ bucket ], variation, data );
+			return { ...productEdits, [ bucket ]: _array };
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js]: ES6 syntax magic :) Nice

File: client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+ */
+import reducer from '../edits-reducer';
+
+import {
+	editProductVariation,
+} from '../actions';
+
+describe( 'edits-reducer', () => {
+	it( 'should initialize to null', () => {
+		expect( reducer( undefined, { type: '@@test/INIT' } ) ).to.equal( null );
```
Comment (allendav) [inline client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js]: I'm not familiar with `@@test/INIT` and a quick google didn't help - can you explain?

File: client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js (hunk 1 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/ui/products/variations/edits-reducer.js]: We started adopting the underscores for the new versions of things that are destined to replace the old ones.

File: client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js (hunk 3 above)
Comment (coderkevin) [inline client/extensions/woocommerce/state/ui/products/variations/test/edits-reducer.js]: It's a throw-away type. It just needs to not be something that our reducer would catch.

**Pull request merged**